src/sdk/namespace_fs.js

```diff
@@ -71,6 +72,74 @@ function make_named_dirent(name) {
     return entry;
 }
 
+/**
+ * @param {fs.Stats} stat 
+ * @returns {string}
+ */
+function get_etag(stat) {
+    // IMPORTANT NOTICE - we must return an etag that contains a dash!
+    // because this is the criteria of S3 SDK to decide if etag represents md5
+    // and perform md5 validation of the data.
+    const ident_str = 'inode-' + stat.ino.toString() + '-mtime-' + stat.mtimeMs.toString();
+    return ident_str;
+}
+
+function check_md_conditions(conditions, stat) {
+    if (!conditions) return;
+    if (!conditions.if_match_etag &&
+        !conditions.if_none_match_etag &&
+        !conditions.if_modified_since &&
+        !conditions.if_unmodified_since) return;
+
+    const data = stat ? {
+        etag: get_etag(stat),
+        last_modified: stat.mtime
+    } : {
+        etag: '',
+        last_modified: 0,
+    };
+
+    // See http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html#req-header-consideration-1
+    // See https://tools.ietf.org/html/rfc7232 (HTTP Conditional Requests)
+    let matched = false;
+    let unmatched = false;
+
+    if (conditions.if_match_etag) {
+        if (!(stat && http_utils.match_etag(conditions.if_match_etag, data.etag))) {
+            throw Object.assign(new Error('check_md_conditions failed'), { rpc_code: 'IF_MATCH_ETAG', rpc_data: data });
+        }
+        matched = true;
+    }
+    if (conditions.if_none_match_etag) {
+        if (stat && http_utils.match_etag(conditions.if_none_match_etag, data.etag)) {
+            throw Object.assign(new Error('check_md_conditions failed'), { rpc_code: 'IF_NONE_MATCH_ETAG', rpc_data: data });
+        }
+        unmatched = true;
+    }
+    if (conditions.if_modified_since) {
+        if (!unmatched && (!stat || conditions.if_modified_since > data.last_modified)) {
+            throw Object.assign(new Error('check_md_conditions failed'), { rpc_code: 'IF_MODIFIED_SINCE', rpc_data: data });
+        }
+    }
+    if (conditions.if_unmodified_since) {
+        if (!matched && (!stat || conditions.if_unmodified_since < data.last_modified)) {
+            throw Object.assign(new Error('check_md_conditions failed'), { rpc_code: 'IF_UNMODIFIED_SINCE', rpc_data: data });
+        }
+    }
+}
+
+function get_transformed_xattr(xattr, add_user_prefix) {
```

Comment:
This should be split to two functions, instead of passing a flag that changes completely what the function does.